### ISSUE TITLE
Reject files based on size on add

### ISSUE THREAD

**Issue by brianmingus** [git-lfs/git-lfs#730]
If you are adding many files and just one of them is > ~2GB and it is near the end, then all of the progress in committing the files is lost as you must remove the file and then start from scratch.

[422] Size must be less than 2147483648

**Comment by brianmingus** [comment]
Ok - it seems that the progress is actually cached. I'll leave this open though since it does seem that git-lfs should notify you of the github size limit when it starts tracking a file.

**Comment by brianmingus** [comment]
As far as I can tell, using git rm --cached and git-lfs untrack on files that are too large does not actually remove them. The total file size to be transmitted continues to reflect all the files that I initially added. In addition to this, the messages telling you that some files are too large do not tell you which files are being referred to.

**Comment by brianmingus** [comment]
I've done everything I can to remote all files from git-lfs. Git reports that my entire data directory is untracked, and git-lfs reports that no files are tracked. Nevertheless, when I try to commit it keeps trying to push all these files, and continues failing because some files are too large.

**Comment by technoweenie** [comment]
It sounds like you still have commits to push that contain large files. You'll have to redo those commits.

> it does seem that git-lfs should notify you of the github size limit when it starts tracking a file.

That'd be nice, but that server side limit has nothing to do with the Git LFS client. We could add an API call every time a file is added, but that would add even more overhead to adding files. I'd be interested in PRs that solve this though.

**Comment by rBatt** [comment]
Issue #1073 says it's an issue with GitHub, and not git lfs. I had not realized such a limitation had been placed on the size of individual files. Either way, that's probably the source of the problem here.

**Comment by otri** [comment]
> That'd be nice, but that server side limit has nothing to do with the Git LFS client. We could add an API call every time a file is added, but that would add even more overhead to adding files. I'd be interested in PRs that solve this though.

I'd argue that it's a core Git LFS usage failure, because the repercussions are non-trivial for all users that encounter it.  A quick usability test of adding a file to any Git LFS repository with a reasonable file size limit causes a forced failure mode that a user must unwind by getting into deep internals of how git functions.  Most clients do not catch this issue, because the error is not brought to the surface.

So we Git LFS must handle this file size restriction through a formalized method.  At bare minimum there should be a way to:
- Determine the file size limits on an associated Git LFS server, and cache this locally
- Warn of adding an over-size file to Git LFS at time of the addition, allow a "--force" option
- Formalize handling of adding a file to the Git LFS server that encounters a restriction, and report the specific file in question
- Allow easy removal of offending file

**Comment by ttaylorr** [comment]
@otri I think that your suggestions are very interesting. I have two questions:

1. What should LFS do when there is a file added to (and tracked by) Git LFS that is too large for it (for example, GitHub limits LFS blobs to 2GiB).
2. What should LFS do when there is a file added to Git that either (a) should be tracked by Git LFS, or (b) is too large for your host and will be rejected during the subsequent `git push`?

I think that (1) is easily addressable, though I'm not sure if it is as useful as addressing (2). The easiest way I can think of to address (2) is to have Git LFS check _all_ files against a `.gitconfig`-urable value, like `lfs.limits.blobsize`, but this relies on the user having installed Git LFS in the first place. So long as Git LFS (and the corresponding hook) is installed correctly, Git LFS can warn about files that (i) don't match any Git LFS include pattern, and (ii) are above the given blob size. I agree that a `--force` option would be useful here as well.

One thing to note, I think that this is tricky to implement in practice at `git-add(1)` time. The earliest hook I can think of would be `pre-commit` (which would give us a `--force`-like with `git commit --no-verify`).

What are your thoughts?

##

/cc @larsxschneider if this is a topic you are interested in working on.

**Comment by otri** [comment]
Without knowing the inner mechanics of `git-add` and `git-lfs` hooks, it's hard for me to say what is the correct place to wire this check in.  Probably a `pre-commit` test like you're saying.

This check would be valid only if user has initialized Git LFS, and has made first contact with the LFS server.  That's the only time I could imagine we could get the `lfs.limits.blobsize` off the server.

My third suggestion above  about `Formalize handling of adding a file to the Git LFS server that encounters a restriction, and report the specific file in question` is actually for handling the case where we don't have a `lfs.limits.blobsize` or if the quota has changed to a smaller value and no longer qualifies, or for when any different kind of LFS server quota / rule kicks in.  Basically, we need to more clearly work out what this error tells the user, and supporting clients like Sourcetree, GitHub, or GitKraken.

Then the fourth suggestion is for providing a user a better means for removing the offending file from a commit.  I discovered that Github has a helpful document for removing the offending file.

[https://help.github.com/articles/removing-files-from-a-repository-s-history/](https://help.github.com/articles/removing-files-from-a-repository-s-history/)

However commands like these `$ git rm --cached giant_file` is entirely too cryptic.  It was actually scary enough that I copied over the entirety of my folder tree before going ahead with the instructions.

For a new user, what does `git rm` actually do?  Does it mean deleting the giant file?  Is it going to delete and also get rid of the cache entry too?  The man page for `git-rm` says:
`Remove files from the index, or from the working tree and the index.`

Definitely not what I would think of as a solution, if I need to preserve a local copy of the giant file in question.

**Comment by larsxschneider** [comment]
@ttaylorr Maybe we don't need to use a `pre-commit` hook here. On `git add` a smudge operation is invoked and at this point Git LFS knows about the file and the size. If a file is >2GB then we could request the max blob size from the server. If the the max blob size is lower than the file we are smudging then we just return an error code to Git with a message. Consequently, the `git add` operation would fail and Git would display the message. Wouldn't that solve the problem? Then users would not need to fuddle with `git rm` and friends.

I think querying the server for every file that is >2GB shouldn't be a problem. Processing/uploading these files takes certainly longer than a single API call. Furthermore, I think it is valid to assume that all LFS implementations support up to 2GB files.

What do you think?

**Comment by ttaylorr** [comment]
> On `git add` a smudge operation is invoked and at this point Git LFS knows about the file and the size.

I'm not sure why I hadn't considered implementing this at the smudge level. I think that is an elegant solution and that is the direction we should take in implementing this.

With regards to checking with the server, I disagree that `git add` should require network connectivity, and would rather implement this by storing a configuration value as above. What do you think?

**Comment by larsxschneider** [comment]
I agree that it is ugly to make network calls in `git add` as this function is supposed to be local only. 

Maybe we hook the "get LFS limit" into `git fetch` then? Whenever we fetch and the variable is not set for the remote yet then we request it from the server. Or maybe we query that variable every time we fetch?

On `git add` we determine the lowest limit for all remotes and use that (users could push code to two different remotes with different Git LFS limits).

**Comment by ttaylorr** [comment]
> Or maybe we query that variable every time we fetch?

I think that this option is fine. It could fit into the Batch API in a new field called `properties`, where the client sends a JSON array of property keys it would like information about, and the server responds with a JSON map of those properties. For instance, the client could send:

```json
{ properties: ["obj.maxsize"] }
```

and the server could respond with:

```json
{ properties: { "obj.maxsize": 2e+9 } }
```

I think that this is tiny enough (and simple to implement) that I don't have an issue with querying it once per `git fetch` or `git push` operation (but not once per batch request).

Perhaps if we don't have this data, we could warn at a pre-determined threshold? What do you think?

**Comment by larsxschneider** [comment]
The `properties` field sounds great 👍 I wonder, how is the "locking support negotiation" implemented? Wouldn't it make sense to have only one way to negotiate the server capabilities? Should we merge these two somehow?

I also wonder if we can make the `properties` fetch happen even on `git clone` (maybe hooked into the `delayed-checkout` code as we have no Git hook in this early stage AFAIK). Then we would only run into the "have no properties data" situation for Git LFS servers that do not support properties. And in this case, I would definitively warn with a pre-determined threshold at 2GB.

**Comment by ttaylorr** [comment]
> I wonder, how is the "locking support negotiation" implemented?

Good question. The way this happens right now is via a separate endpoint, where the client sends the locks that it wants, and the server sends the list of locks that it can obtain. More about that in detail is [here](https://github.com/git-lfs/git-lfs/blob/master/docs/api/locking.md).

I think that I'd prefer for this addition to not also have it inhabit lock verification. Doing so would mean that implementors of the specification would also have to rework their implementation of locking through this new API, which I'd like to avoid now if possible.

I think that a good time to revisit this in the future would be if the locking API were to change, we could ask how it should fit into this new properties field.

> I also wonder if we can make the `properties` fetch happen even on `git clone` (maybe hooked into the `delayed-checkout` code as we have no Git hook in this early stage AFAIK). Then we would only run into the "have no properties data" situation for Git LFS servers that do not support properties. And in this case, I would definitively warn with a pre-determined threshold at 2G

That sounds good to me.

##

Speaking more broadly, would you be interested in opening a proposal issue so that we could discuss this more formally?

**Comment by serbinsh** [comment]
I dont see any updates since Feb, has this been addressed or finalized?  is there a beta option we can try?  Or is there a timeline to support this?

Thanks!

**Comment by bk2204** [comment]
This issue hasn't been implemented yet and we don't currently have it planned.  That doesn't mean that it couldn't be implemented, only that we're busy and have more things to implement than time to implement them.

I think we're currently waiting for a formal proposal to be drawn up before implementing it, so perhaps @larsxschneider would be interested in opening an issue or PR for that purpose?

**Comment by mbcrump** [comment]
Any updates on this? We just hit this on our end. // @bk2204 or @larsxschneider

**Comment by bk2204** [comment]
Nope, we don't have any updates on this.  If someone wanted to write up a proposal, we'd be happy to take a look and provide feedback.

**Comment by KenHenson** [comment]
An easy fix is to find the offending file in your project, remove it or break it up into smaller files and resubmit then entire project to a clean repository.

**Comment by otri** [comment]
I'll try to paint a motivating picture as to why this edge case is important to fully solve.

This file upload storage limit situations is one of those awful edge cases that the average software engineer rarely would ever hit.  However once a project involves any artists committing large assets (ie game developers with artists) this gets hit fairly readily.  In this case the software engineer gets yanked out of what they are working on and has to assist the artist.  This issue creates friction and is complicated enough that it can take over a full day to get educated enough to solve properly, sucking up at least two people's time.  So it makes git-lfs a bit difficult for many indie game developers, because it's unresolvable and an education thing for every artist. 

It's the sort of things that leave a lasting bitter experience, and motivates studios to use products like Perforce despite the cost.  And ideally it's something that git-lfs would greatly benefit from to make it "just work™" and if not, then make it clear why something can't work before it becomes a "huge pain™".  Removing an LFS file from commit history and pushing a re-written history while in active concurrent development with other commits, is definitely classed as the latter.

**Comment by jacobmott** [comment]
If im paying for it(Im paying $15 a month on github for 150GB storage), why wouldn't you support it? Why do you care why i want to store something larger than 2gb? (In my case its a Unreal engine marketplace texture/megascans pak used in my game im creating, impossible to break up into pieces)? if im paying you why do you care.. im specifically paying you for that space/size? Now i have to write a script to upload to an aws s3 bucket and download when i checkout the repo.. Nice job. your just making peoples lives harder, and not really accomplishing anything in the process, other than making things un-optimal for people. If your trying to make things hard/unmaintainable/unreasonable/unoptimized you've done it, that's not the way i solution things. Or is there a legitimate reason for the 2GB reason other than, o we just don't think you should be storing files that large(Basically your personal/biased opinions) ?

**Comment by otri** [comment]
@jacobmott your situation is almost exactly what I was trying to paint a picture about earlier.

Is this the right ticket?  Yes, and no, this is the right ticket to vent about how difficult the thing was to deal with.  But the 2GB per file limit is probably a github LFS repo limitation (it used to be 1GB per file limit).

In any case, there's got to be a better way for this to work for game development.  And actually, I'm bumping into this with larger AI models involving 3D volumes and lightfields too.  These iteratively improved models definitely merit some kind of version control and GIT LFS, with unspoken limitations held on the server, is falling behind in this scenario as well.

**Comment by bk2204** [comment]
Hey, folks.

Git LFS handles arbitrary files up to 2^63-1 bytes in size, although there are some practical concerns around the performance of hashing very large files.  This repo handles the open source client, which anyone can use and modify.  If you have complaints about GitHub's (or GitLab's, or Bitbucket's) implementation of the Git LFS server, please let them know.  Telling us here isn't going to result in any changes in that regard.
